homeassistant/components/switcher_kis/__init__.py

```diff
@@ -0,0 +1,93 @@
+"""Home Assistant Switcher Component."""
+
+from asyncio import TimeoutError as Asyncio_TimeoutError, wait_for
+from datetime import datetime, timedelta
+from logging import getLogger
+from typing import Dict, Optional
+
+import voluptuous as vol
+
+from homeassistant.components.switch import DOMAIN as SWITCH_DOMAIN
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.discovery import async_load_platform
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import EventType, HomeAssistantType
+
+_LOGGER = getLogger(__name__)
+
+DOMAIN = 'switcher_kis'
+
+CONF_DEVICE_ID = 'device_id'
+CONF_DEVICE_PASSWORD = 'device_password'
+CONF_PHONE_ID = 'phone_id'
+
+DATA_DEVICE = 'device'
+
+SIGNAL_SWITCHER_DEVICE_UPDATE = 'switcher_device_update'
+
+ATTR_AUTO_OFF_SET = 'auto_off_set'
+ATTR_DEVICE_NAME = 'device_name'
+ATTR_ELECTRIC_CURRNET = 'electric_current'
+ATTR_IP_ADDRESS = 'ip_address'
+ATTR_LAST_DATA_UPDATE = 'last_data_update'
+ATTR_LAST_STATE_CHANGE = 'last_state_change'
+ATTR_REMAINING_TIME = 'remaining_time'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_PHONE_ID): cv.string,
+        vol.Required(CONF_DEVICE_ID): cv.string,
+        vol.Required(CONF_DEVICE_PASSWORD): cv.string
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass: HomeAssistantType, config: Dict) -> bool:
+    """Set up the switcher component."""
+    from aioswitcher.bridge import SwitcherV2Bridge
+
+    phone_id = config[DOMAIN][CONF_PHONE_ID]
+    device_id = config[DOMAIN][CONF_DEVICE_ID]
+    device_password = config[DOMAIN][CONF_DEVICE_PASSWORD]
+
+    v2bridge = SwitcherV2Bridge(
+        hass.loop, phone_id, device_id, device_password)
+
+    await v2bridge.start()
+
+    async def async_stop_bridge(event: EventType) -> None:
+        """On homeassistant stop, gracefully stop the bridge if running."""
+        await v2bridge.stop()
+
+    hass.async_add_job(hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STOP, async_stop_bridge))
+
+    try:
+        device_data = await wait_for(
+            v2bridge.queue.get(), timeout=5.0, loop=hass.loop)
+    except (Asyncio_TimeoutError, RuntimeError):
+        _LOGGER.exception("failed to get response from device")
+        await v2bridge.stop()
+        return False
+
+    hass.data[DOMAIN] = {
+        DATA_DEVICE: device_data
+    }
+
+    hass.async_create_task(async_load_platform(
+        hass, SWITCH_DOMAIN, DOMAIN, None, config))
+
+    async def device_updates(timestamp: Optional[datetime]) -> None:
+        """Use for updating the device data from the queue."""
+        if v2bridge.running:
+            device_new_data = await v2bridge.queue.get()
```

Comment:
@andrewsayre 
I had thought about it actually, but I'm not sure what else can I do at this point other then await for a message to be produced into the queue.
Any ideas?

Thank you.

Reply:
I think you need timeout so the intervals don't back up.  Alternatively, the library could take in a `callable` during `__init__` that it invokes with the data when it arrives, basically a simplified dispatcher.  That's the pattern other libraries use when data is pushed and not polled.

Reply:
@andrewsayre 
I'm sorry friend, I might need a little clarification on this one.
If you remember the previous PR.
Originally the library's `__init__` took a `callable` as a callback argument and pushed its data into it.
That was replaced with `asyncio.Queue` for a more friendly and non-blocking design.

Isn't `asyncio.Queue` does exactly what we're trying to accomplish here?
When empty, it should be awaited non-blocking until a message has been produced into the queue.
And if so... is this change request a "deal-breaker"?
Because the library is undergoing some changes these days and I can't actually revert my current work and I'm afraid requesting a change in the library will block the progress here for a while and that's a shame.
:-(

Reply:
Lib changes aren't necessary - just offering how others have solved it.  So I think the best way to solve this would be to use a background task that runs in a loop awaiting the queue.  You'd then cancel this task on stop:
```python

async def queue_monitor():
    while v2bridge.running:
        device_new_data = await v2bridge.queue.get()
        ...

monitor_task = hass.async_create_task(queue_monitor())
```

To stop it:
```python
monitor_task.cancel()
try:
    await monitor_task()
catch asyncio.CancelledError:
    pass
```

Reply:
Got it, Thanks.
I'll get right on it.

Reply:
@andrewsayre 
I think I'm too tired (it's 1 am here :-) ),
I just understood your original change request.

The device is suppose to add a data object to the queue every 4 seconds.
I think I can simply schedule the `async_track_time_interval` to run every 4 seconds instead of 3,
and on the function end I'll change the `queue.get()` to `queue.get_nowait()`.
I believe this should resolve the potential issue of backing up the update intervals.
Something like this:
```python
    async def device_updates(timestamp: Optional[datetime]) -> None:
        """Use for updating the device data from the queue."""
        if v2bridge.running:
            try:
                device_new_data = v2bridge.queue.get_nowait()
                if device_new_data:
                    async_dispatcher_send(hass,
                                          SIGNAL_SWITCHER_DEVICE_UPDATE,
                                          device_new_data)
            except QueueEmpty:
                pass

    async_track_time_interval(hass, device_updates, timedelta(seconds=4))
```

Any thoughts?

Reply:
That's fine - its unlikely the two timers will be synchronized, so I would expect there to be a 1-4s delay for HASS to update when new data is available.  Switching to the task approach would resolve that, but we can tackle that in the future.